### ISSUE TITLE
Grails 7: `render(text: "...")` not working

### ISSUE THREAD

**Issue by matrei** [codeconsole/grails-sitemesh3#4]
When using `render(text: "...")` in a controller, Sitemesh tries to find a layout and renders a 404:

```console
2024-11-18T11:52:23.494+01:00 ERROR 743942 --- [o-auto-1-exec-8] org.sitemesh.webapp.WebAppContext        : Error 404 processing decorator '/layouts/_none_'
```

See example project with tests: https://github.com/matrei/sitemesh-bug-render-text

**Comment by codeconsole** [comment]
@matrei it sounds like the issue here isn't `render(text:)` it is that you are setting the contentType to `text/html` and sitemesh is configured to decorate all `text/html` content.   Can you confirm if removing `contentType` fixes the issue?

Does setting "text/plain" work?

**Comment by jdaugherty** [comment]
Isn't it valid to render adhoc html templates (i.e. for htmx)?  I don't think sitemesh should assume it needs to decorate if there's not a layout.

**Comment by codeconsole** [comment]
> If a layout is not located using any of those conventions, the convention of last resort is to look for the application default layout which is grails-app/views/layouts/application.gsp.

https://gsp.grails.org/latest/guide/layouts.html

**Comment by jdaugherty** [comment]
I would only expect this behavior if a view is being rendered.  

In grails 6.2.x, if you have this controller: 

    class TestingController {
	    def index() {
		    render text: '<html><body>Testing</body></html>', contentType: 'text/html'
	    }
    }

It renders `<html><body>Testing</body></html>` and does not use the default layout.  In 7.x with this plugin, it's trying to render a layout - even though one doesn't exist.

**Comment by codeconsole** [comment]
@jdaugherty IMO that is a view.  If I did that, I would want it to decorate.  

If you are rendering html, why would you not want it to decorate if you have default layouts enabled?

**Comment by jdaugherty** [comment]
Because not everyone uses GSPs.  A good example of this is if you wnat to render a single form element - for cases like htmx or where you want to render an html fragment.  I would have never expected sitemesh to try to decorate unless render view: <value> was called.

**Comment by jdaugherty** [comment]
I guess the other question is: does sitemesh also decorate templates now?

**Comment by codeconsole** [comment]
no, it does not decorate templates.

**Comment by matrei** [comment]
> Does setting "text/plain" work?

@codeconsole Yes, that works.
I have updated https://github.com/matrei/sitemesh-bug-render-text with more tests to show that.

However, `render(text: '<p>Hello World</p>', contentType: 'text/html')` has been working for previous versions of Grails in `grails-functional-tests` since 2015.

https://github.com/grails/grails-functional-tests/blob/2df395bf60d42c4d596837453dc717e8c3d638e4/app1/grails-app/controllers/functionaltests/UploadController.groovy#L14

**Comment by codeconsole** [comment]
well, an easy solution for this would be to not have a default decorator.   In sitemesh 3 you can decorate anything, including static html pages. 

How do you guys feel about eliminating the default decorator?

**Comment by matrei** [comment]
>well, an easy solution for this would be to not have a default decorator.

Would that be backwards compatible with how Grails < 7 works?

**Comment by codeconsole** [comment]
> > well, an easy solution for this would be to not have a default decorator.
> 
> Would that be backwards compatible with how Grails < 7 works?

no, but making everything work how it worked previously isn't necessarily a good thing.   The correct decision is figuring out which is the best way it should work.   Having a default decorator, from what I remember, is inefficient because it assumes every html page will be decorated.

**Comment by jdaugherty** [comment]
I'm not as familiar with the updated sitemesh.  Do you have any documentation on the default decorator that I can refer to?  I assume the idea is you'd have to include the layout in the gsp if you want it decorated or in the render command?  

In general, I'm not a fan of decorating explicit render(text: 'text here') calls from the controller.  If someone says I want to render some text, I'd only ever expected it rendered.  If they want to add the layout, they can always call a view or explicitly set the layout argument.

Another possible solution is to only apply the layout when it's a view, or if the layout: '' parameter is set.  This may be the closest to pre Grails 7 functionality and would make the most sense in my mind.

**Comment by codeconsole** [comment]
I misread the original topic because the error scrolled off the page and I assumed "_tries to find a layout_" meaning it is looking for the default layout (this is currently possible but a opt in configuration).  Looking more at the actual error message:

```
Error 404 processing decorator '/layouts/_none_'
```

I don't think the issue here is with the plugin. It is the fact that a layout is being set to `_none_` and site mesh has no concept of a none layout.  If the layout was not set to none, there would be no error.

https://github.com/grails/grails-core/blob/cc5d4cae8d8453824849957caef9ae6635b2a52b/grails-web-common/src/main/groovy/org/grails/web/util/WebUtils.java#L64

https://github.com/grails/grails-core/blob/cc5d4cae8d8453824849957caef9ae6635b2a52b/grails-plugin-controllers/src/main/groovy/grails/artefact/controller/support/ResponseRenderer.groovy#L535-L544

A simple fix is just not setting the layout, instead of setting it to none.   A more involved fix would be allowing layout overriding and implementing a none layout to achieve that goal.

**Comment by codeconsole** [comment]
I think we should consider removing all sitemesh references in ResponseRenderer. I am not sure if it makes sense to compile sitemesh logic into RestConteoller and Interceptor

**Comment by matrei** [comment]
Fixed by grails/grails-core#13908
